compiler/noirc_evaluator/src/ssa/opt/die.rs

```diff
@@ -390,7 +391,14 @@ fn can_be_eliminated_if_unused(
             Value::ForeignFunction(_) => false,
 
             // We must assume that functions contain a side effect as we cannot inspect more deeply.
-            Value::Function(_) => false,
+            // We use purity to determine whether functions contain side effects.
+            // If we have an impure function, we cannot remove it even if it is unused.
+            Value::Function(function_id) => match function.dfg.purity_of(function_id) {
+                Some(Purity::Pure) => true,
+                Some(Purity::PureWithPredicate) => true,
```

Comment:
Is this right? Seems like this case should depend on the predicate value. E.g. a function with a `constrain` inside will be marked as this and we don't want to remove those unless we know the predicate is false.

Reply:
Yeah you are correct. I just wanted to see the effects here. I hadn't looked deeply into the purity types but I see now that we mark functions with `constrain` as being `PureWithPredicate`. I am most likely going to add some more granular purity types.

Reply:
What would the more granular cases be for?

Reply:
I was thinking along the lines of that handling `constrain` and binary ops are different depending on whether we want to deduplicate, hoist, or eliminate. However, after looking a bit deeper we probably just want to not eliminate checked binary ops at all as suggested by @TomAFrench. 

For now, I think I will just switch `PureWithPredicate` to `false`. We can add checking whether the predicate is false in a separate PR, as we will need to handle checking both control dependence for Brillig and the side effects var for ACIR.

Reply:
Tracking here https://github.com/noir-lang/noir/issues/8302